projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-attach/attach.cpp

```diff
@@ -0,0 +1,90 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "attach.h"
+#include "code_object_registration.hpp"
+#include "queue_registration.hpp"
+#include "table.hpp"
+
+#include "lib/common/logging.hpp"
+
+void
+init_logging()
+{
+    rocprofiler::common::init_logging("ROCPROFILER_ATTACH");
+}
+
+// ensure that logging is always initialized when library is loaded
+bool init_logging_at_load = (init_logging(), true);
+
+using rocprofiler_register_functor_t = void (*)(uint64_t, void*, uint64_t);
+
+ROCPROFILER_EXTERN_C_INIT
+
+int
+rocprofiler_attach_set_api_table(const char* name,
+                                 uint64_t    lib_version,
+                                 uint64_t    lib_instance,
+                                 void**      tables,
+                                 uint64_t    num_tables,
+                                 void*       register_functor)
+{
+    ROCP_TRACE << "rocprofiler_attach_set_api_table called for api " << name;
+    (void) lib_version;   // unused
+    (void) lib_instance;  // unused
+
+    if(std::string_view{name} != "hsa")
+    {
+        ROCP_ERROR << "rocprofiler_attach_set_api_table was called with a table other than HSA";
+        return ROCPROFILER_STATUS_ERROR_INVALID_ARGUMENT;
+    }
+
+    ROCP_ERROR_IF(num_tables > 1) << "rocprofiler expected HSA library to pass 1 API table, not "
+                                  << num_tables;
+
+    auto* hsa_api_table = static_cast<HsaApiTable*>(tables[0]);
+
+    rocprofiler::attach::dispatch_table_init();
+
+    if(register_functor)
+    {
+        auto callable = reinterpret_cast<rocprofiler_register_functor_t>(register_functor);
+        callable(rocprofiler_attach_get_version(),
+                 (void*) (rocprofiler::attach::get_dispatch_registration_table()),
+                 1);
+    }
+
+    // Initialize all registration services in attach
+    rocprofiler::attach::queue_registration_init(hsa_api_table);
+    rocprofiler::attach::code_object_registration_init(hsa_api_table);
+
+    return ROCPROFILER_STATUS_SUCCESS;
+}
+
+int
+rocprofiler_attach_get_version()
+{
+    constexpr int ROCPROFILER_ATTACH_VERSION = 1;
+    return ROCPROFILER_ATTACH_VERSION;
+}
```

Comment:
Why only a single version number? This needs to be fixed to be the same form as `ROCPROFILER_VERSION` which encodes the major, minor, and patch